controller/websocket_controller.go

```diff
@@ -111,8 +104,14 @@ func (wc *WebsocketController) Run(stopCh <-chan struct{}) {
 }
 
 func (wc *WebsocketController) Close() {
-	for _, w := range wc.watchers {
-		w.Close()
+	// Since the Close function on the Watcher modifies the slice in place, we
+	// could have problems if iterating over it in order (one item gets removed,
+	// next item gets pushed down to old index and skipped, etc.). To safely
+	// remove the items from the slice as we're doing now, we need to remove them
+	// from the end, which means we need to loop through our slice backwards.
+	for i := len(wc.watchers) - 1; i >= 0; i-- {
```

Comment:
The lock is there to protect wc.watchers, not the single watcher.

You don't need to do it in reverse order now.

It should work if you just grab the lock, close everything in order in the loop, clean up the wc.watchers, and release the lock.